packages/cogify/src/cogify/cli/cli.cog.ts

```diff
@@ -49,150 +64,108 @@ export const BasemapsCogifyCreateCommand = command({
-      const cogAsset = item.assets['cog'];
+      const cogAsset = f.item.assets['cog'];
       if (cogAsset == null) return true;
-
-      // Force overwrite existing files
+      // Force overwrite existing COGs
       if (args.force) {
-        logger.info({ asset: cogAsset.href }, 'Cog:Create:Overwrite');
+        logger.warn({ item: f.item.id, asset: cogAsset.href }, 'Cog:Create:Overwrite');
         return true;
       }
 
-      logger.info({ asset: cogAsset.href }, 'Cog:Create:Exists');
-      // TODO check if the tiff file actually exists
-      // TODO validate checksum of tiff file?
+      logger.info({ item: f.item.id, asset: cogAsset.href }, 'Cog:Create:Exists');
       return false;
-    });
+    }) as CogItem[];
 
-    // Nothing to create exit early
-    if (toCreate.length === 0) {
-      logger.info('Cog:Exists');
+    // No items left to be created
+    if (filtered.length === 0) {
+      logger.info({ toCreate: filtered.length }, 'Cog:Create:Done');
       return;
     }
+    const tmpFolder = fsa.join(tmpdir(), CliId);
 
     // Get list of unique source files needed for all files
-    const sourceFiles = new Set<string>();
-    for (const i of toCreate) {
-      const files = getSources(i.links);
-      for (const src of files) sourceFiles.add(src.href);
-
-      const cutline = getCutline(i.links);
-      if (cutline) sourceFiles.add(cutline.href);
+    const sources = new SourceDownloader(tmpFolder);
+    for (const i of filtered) {
+      const files = getSources(i.item.links);
+      for (const src of files) sources.register(src.href, i.item.id);
     }
 
-    // Mapping of remote filename to source filename
-    const sourceFileMap = new Map<string, string>();
-    const tmpFolder = fsa.join(tmpdir(), CliId);
-
     const gdalVersion = await new GdalRunner({ command: 'gdal_translate', args: ['--version'], output: '' }).run();
 
     try {
       await mkdir(tmpFolder, { recursive: true });
 
-      await Promise.all(
-        [...sourceFiles].map(async (f) => {
-          // No need to download files that exist locally
-          if (isPathLocal(f)) {
-            sourceFileMap.set(f, f);
-            return;
-          }
-          // Create a temporary file name that is the base58 encoded path, this allows us to
-          const newFileName = sha256base58(Buffer.from(f)) + extname(f);
-          const targetFile = fsa.joinAll(tmpFolder, 'source', newFileName);
-
-          logger.debug({ source: f, target: targetFile }, 'Cog:Source:Download');
-          const startTime = performance.now();
-          await fsa.write(targetFile, fsa.stream(f));
-          const duration = performance.now() - startTime;
-          logger.debug({ source: f, target: targetFile, duration }, 'Cog:Source:Download:Done');
-          sourceFileMap.set(f, targetFile);
-        }),
-      );
-
-      for (const item of toCreate) {
+      // TODO should COG creation be run concurrently?
+      for (const { itemPath, item } of filtered) {
```

Comment:
I think we already allow to running concurrent gdal commands previously.

Reply:
It is pretty easy to configure into concurrent execution, but unsure if its worth the extra complexity in this pull request.

I could look to add it in after some initial performance tests in Argo?